src/net.cpp

```diff
@@ -1924,16 +1922,28 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
         int nTries = 0;
         while (!interruptNet)
         {
-            CAddrInfo addr = addrman.SelectTriedCollision();
-
-            // SelectTriedCollision returns an invalid address if it is empty.
-            if (!fFeeler || !addr.IsValid()) {
-                addr = addrman.Select(fFeeler);
-            }
-
-            // Require outbound connections, other than feelers, to be to distinct network groups
-            if (!fFeeler && setConnected.count(addr.GetGroup(addrman.m_asmap))) {
-                break;
+            CAddrInfo addr;
+
+            // For non-feeler connections, attempt to connect to an address among the collisions
+            // in our "tried" table.
+            // For feelers iterations, since feelers should be selected from the "new" table in the AddrMan,
+            // skip selecting collisions.
+            // Also require non-feelers to be distinct network groups to preserve connection diversity.
+            // It's much less relevant for feelers, because it's a frequent procedure
+            // to sanitize AddrMan (not critical), and selecting from AddrMan is
+            // sufficiently (for this case) random/difficult-to-Sybil due to the bucketing.
+            if (conn_type != ConnectionType::FEELER) {
```

Comment:
It's rare that we get to this line of code -- on startup, when we have not enough outbound peers, we'll hit this.  But once we have made our initial 10 outbound connections, then at line 1916 above we'll abort the loop iteration, never getting to this code.

So what would happen is that our FEELERS generate tried-table-collisions which never get resolved, eventually resulting in the old tried table entry getting evicted without an attempt having been made, breaking test-before-evict.

Reply:
Also, making the test-before-evict connections no longer be FEELER breaks the logic in `net_processing` around disconnecting these peers right when we connect.